### PR TITLE
BAVL-123 changes to check the location for a booking is a valid location at a given prison.

### PR DESCRIPTION
**Pull request by matt-crowson-moj**
These changes now mean to create a booking you'll need to use a valid location key.

This will mean the prison code will also need match that in the location key e.g. `prison code = ABC` and `location key = ABC-1-23456`

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/client/locationsinsideprison/LocationsInsidePrisonClient.kt [hunk 1]
```diff
@@ -19,3 +22,41 @@ class LocationsInsidePrisonClient(private val locationsInsidePrisonApiWebClient:
     .onErrorResume(WebClientResponseException.NotFound::class.java) { Mono.empty() }
     .block()
 }
+
+@Component
+class LocationValidator(private val locationsInsidePrisonClient: LocationsInsidePrisonClient) {
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/client/locationsinsideprison/LocationsInsidePrisonClient.kt:27]: So not 100% convinced on this but I thought I'd try a slim down the validation login the in the creation service.  It is likely similar validation will be required for edit/update.  Keeps the testing simpler in there services as well.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/client/prisonersearch/PrisonerSearchClient.kt [hunk 2]
```diff
@@ -17,10 +17,13 @@ class PrisonerSearchClient(private val prisonerSearchApiWebClient: WebClient) {
...
+}
 
-  fun getPrisonerAtPrison(prisonerNumber: String, prisonCode: String): Prisoner =
-    getPrisoner(prisonerNumber)?.takeUnless { prisoner -> prisoner.prisonId != prisonCode }
-      ?: throw EntityNotFoundException("Prisoner $prisonerNumber not found at prison $prisonCode")
+@Component
+class PrisonerValidator(val prisonerSearchClient: PrisonerSearchClient) {
+  fun validatePrisonerAtPrison(prisonerNumber: String, prisonerCode: String): Prisoner =
+    prisonerSearchClient.getPrisoner(prisonerNumber)?.takeUnless { prisoner -> prisoner.prisonId != prisonerCode }
+      ?: throw ValidationException("Prisoner $prisonerNumber not found at prison $prisonerCode")
```
Comment (timharrison-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/client/prisonersearch/PrisonerSearchClient.kt:26]: Should `prisonerCode` be `prisonCode` ?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/client/locationsinsideprison/LocationsInsidePrisonClient.kt [hunk 3]
```diff
@@ -19,3 +22,41 @@ class LocationsInsidePrisonClient(private val locationsInsidePrisonApiWebClient:
...
+
+  fun validatePrisonLocation(prisonCode: String, locationKey: String) {
+    validatePrisonLocations(prisonCode, setOf(locationKey))
+  }
```
Comment (timharrison-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/client/locationsinsideprison/LocationsInsidePrisonClient.kt:31]: What happens if the person is temporarily out of the prison when the appointment is booked? e.g. he's on day-release or job ROTL? Not sure what prisoner search will tell us where they are?

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/client/prisonersearch/PrisonerValidatorTest.kt [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+  @Test
+  fun `should fail validation when prisoner found but not at prison`() {
+    prisoner.stub { on { prisonId } doReturn MOORLAND }
+
+    whenever(client.getPrisoner("123456")) doReturn prisoner
+
+    val error = assertThrows<ValidationException> { validator.validatePrisonerAtPrison("123456", BIRMINGHAM) }
+
+    error.message isEqualTo "Prisoner 123456 not found at prison BMI"
```
Comment (timharrison-moj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/client/prisonersearch/PrisonerValidatorTest.kt:39]: Might be OUT, but only temporarily?

**Review by timharrison-moj — APPROVED** [review]
LGTM.
Couple of questions added.
Have the location and prisoner search API clients been wrapped in a validator? Is that as clear as a standalone validating which takes results from the API clients and operates on them?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/client/locationsinsideprison/LocationsInsidePrisonClient.kt (hunk 3 above)
Comment (matt-crowson-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/client/locationsinsideprison/LocationsInsidePrisonClient.kt:31]: Erm this is validating a location, did you mean that for here???

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/client/prisonersearch/PrisonerSearchClient.kt (hunk 2 above)
Comment (matt-crowson-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/client/prisonersearch/PrisonerSearchClient.kt:26]: yes it should be `prisonCode`.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/client/prisonersearch/PrisonerValidatorTest.kt (hunk 4 above)
Comment (matt-crowson-moj) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/client/prisonersearch/PrisonerValidatorTest.kt:39]: So I have not really gone to far with validating a prisoners whereabouts.  It gets very complicated and we probably want to decide how far we need to go.

All we are saying at the moment is the prisoner at the prison.  They can still be active in or active out here.  We are not asking if they are in the prison.

Comment (timharrison-moj) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/client/prisonersearch/PrisonerValidatorTest.kt:39]: OK sounds good. I was thinking we just need an indication they are at the prison, and that the apppointment location is valid for the prison, and not much more beyond that.

**Pull request merged**